File: providers/journal/test/journal_test.exs
defmodule JournalTest do
  use ExUnit.Case

  alias Journal.{Command, Parse}

  @alice String.duplicate("a1", 16)
  @bob String.duplicate("b2", 16)

  setup do
    root = Path.join(System.tmp_dir!(), "journal-test-#{System.unique_integer([:positive])}")
    :ok = Journal.Store.ensure(root)
    on_exit(fn -> File.rm_rf!(root) end)
    {:ok, root: root}
  end

  defp run(root, from, line), do: Command.run(root, from, line)

  test "parse handles quotes inside option values" do
    {args, opts} =
      Parse.parse(~s(write hrs/nb/p --body "He said "hi" today" --title "T" --if-rev ""))

    assert args == ["write", "hrs/nb/p"]
    assert opts["body"] == ~s(He said "hi" today)
    assert opts["title"] == "T"
    refute Map.has_key?(opts, "if_rev")
  end

  test "parse bare flags and positional quotes" do
    {args, opts} = Parse.parse(~s(search "learning rate" --project hrs --deep))
    assert args == ["search", "learning rate"]
    assert opts == %{"project" => "hrs", "deep" => true}
  end

  test "write, read, ls, rev and conflict", %{root: root} do
    assert {:ok, "rev: " <> rev} =
             run(root, @alice, ~s(write hrs/ab/p1 --title "P one" --body "# Hello\\n\\nworld"))

    assert {:ok, out} = run(root, @alice, "read hrs/ab/p1")
    assert out =~ "rev: #{rev}"
    assert out =~ ~s(title: "P one")
    assert out =~ "author: #{@alice}"
    assert out =~ "# Hello\n\nworld"

    assert {:ok, "hrs/ab/p1\tP one"} = run(root, @alice, "ls hrs/ab")
    assert {:ok, "hrs"} = run(root, @alice, "ls")

    assert {:error, "conflict" <> _} =
             run(root, @alice, ~s(write hrs/ab/p1 --body "x" --if-rev deadbee))

    assert {:ok, _} = run(root, @alice, ~s(write hrs/ab/p1 --body "x" --if-rev #{rev}))
  end

  test "acl: creator owns, others forbidden until added", %{root: root} do
    {:ok, _} = run(root, @alice, ~s(write hrs/ab/p1 --body "x"))
    assert {:error, "forbidden"} = run(root, @bob, "read hrs/ab/p1")
    assert {:error, "forbidden"} = run(root, @bob, ~s(write hrs/ab/p2 --body "y"))
    assert {:error, "forbidden" <> _} = run(root, @bob, "acl hrs add #{@bob}")
    assert {:ok, _} = run(root, @alice, "acl hrs add #{@bob}")
    assert {:ok, _} = run(root, @bob, "read hrs/ab/p1")
    assert {:ok, "hrs"} = run(root, @bob, "ls")
  end

  test "append, edit and history", %{root: root} do
    {:ok, "rev: " <> _} = run(root, @alice, ~s(write hrs/ab/p1 --body "line one"))
    {:ok, "rev: " <> rev} = run(root, @alice, ~s(append hrs/ab/p1 "line two, quoted"))
    {:ok, out} = run(root, @alice, "read hrs/ab/p1")
    assert out =~ "line one\n\nline two, quoted\n"

    assert {:error, "missing --if-rev"} =
             run(root, @alice, ~s(edit hrs/ab/p1 --find one --replace uno))

    {:ok, "rev: " <> _} =
      run(root, @alice, ~s(edit hrs/ab/p1 --if-rev #{rev} --find "line one" --replace "line uno"))

    {:ok, out} = run(root, @alice, "read hrs/ab/p1")
    assert out =~ "line uno"

    {:ok, hist} = run(root, @alice, "history hrs/ab/p1")
    assert length(String.split(hist, "\n")) == 3
    assert hist =~ String.slice(@alice, 0, 12)
  end

  test "kpi set, log, latest and page injection", %{root: root} do
    {:ok, _} = run(root, @alice, ~s(write hrs/ab/p1 --body "AUC: <!-- kpi: auc -->"))
    {:ok, _} = run(root, @alice, ~s(kpi set hrs/ab auc 0.85 --ref abc123))
    {:ok, _} = run(root, @alice, ~s(kpi set hrs/ab auc 0.871 --note "lr=3e-4"))
    {:ok, _} = run(root, @alice, ~s(kpi set hrs/ab loss 12))

    {:ok, log} = run(root, @alice, "kpi log hrs/ab auc")
    assert [l1, l2] = String.split(log, "\n")
    assert l1 =~ "auc=0.85" and l1 =~ "ref=abc123"
    assert l2 =~ "auc=0.871" and l2 =~ "note=lr=3e-4"

    {:ok, latest} = run(root, @alice, "kpi latest hrs/ab")
    assert latest =~ "auc=0.871"
    assert latest =~ "loss=12"

    {:ok, page} = run(root, @alice, "read hrs/ab/p1")
    assert page =~ "AUC: auc = 0.871 ("
    assert {:error, "invalid_number"} = run(root, @alice, "kpi set hrs/ab auc high")
  end

  test "attach, fetch, link", %{root: root} do
    {:ok, _} = run(root, @alice, ~s(write hrs/ab/p1 --body "x"))
    b64 = Base.encode64("hello bytes")
    {:ok, out} = run(root, @alice, ~s(attach hrs/ab/p1 --name hello.txt --base64 "#{b64}"))
    [_, "sha256: " <> sha] = String.split(out, "\n")
    assert {:ok, ^b64} = run(root, @alice, "fetch #{sha}")

    {:ok, _} = run(root, @alice, ~s(link hrs/ab/p1 file:///Users/zim/runs/x --name run-dir))
    {:ok, _} = run(root, @alice, ~s(link hrs/ab/p1 s3://bucket/ckpt.pt --sha256 abc))
    {:ok, page} = run(root, @alice, "read hrs/ab/p1")
    assert page =~ "name: hello.txt"
    assert page =~ "sha256: #{sha}"
    assert page =~ "uri: file:///Users/zim/runs/x"
    assert page =~ "host: #{@alice}"
    assert page =~ "uri: s3://bucket/ckpt.pt"

    big = Base.encode64(:binary.copy(<<0>>, Journal.Config.max_blob_bytes() + 1))

    assert {:error, "too_large" <> _} =
             run(root, @alice, ~s(attach hrs/ab/p1 --name big --base64 "#{big}"))
  end

  test "read with line range", %{root: root} do
    {:ok, _} = run(root, @alice, ~s(write hrs/ab/p1 --body "a\\nb\\nc\\nd"))
    {:ok, out} = run(root, @alice, "read hrs/ab/p1 --lines 6:7")
    assert out =~ ~r/^rev: [0-9a-f]+\n# ?a\nb$/ or out =~ ~r/^rev: [0-9a-f]+\na\nb$/
  end

  test "rejects bad addresses and missing key", %{root: root} do
    assert {:error, "invalid_address"} = run(root, @alice, ~s(write hrs/ab --body x))
    assert {:error, "invalid_address"} = run(root, @alice, ~s(write ../x/y --body x))
    assert {:error, "forbidden" <> _} = run(root, "", "ls")
  end

  test "stdio protocol round trip", %{root: root} do
    import ExUnit.CaptureIO

    out =
      capture_io(fn ->
        Journal.Stdio.handle_line(
          root,
          ~s({"op":"request","message":"write hrs/ab/p1 --body \\"hi\\"","from":"#{@alice}","request_id":"r1"})
        )

        Journal.Stdio.handle_line(
          root,
          ~s({"op":"request","message":"read hrs/ab/p1","from":"#{@bob}","request_id":"r2"})
        )
      end)

    [l1, l2] = String.split(String.trim(out), "\n")
    assert %{"op" => "reply", "request_id" => "r1", "reply" => "rev: " <> _} = :json.decode(l1)

    assert %{"op" => "error", "request_id" => "r2", "error" => "forbidden"} =
             :json.decode(l2)
  end
end

File: cmd/journal-provider/journal_test.go
package main

import (
	"encoding/base64"
	"os"
	"strings"
	"testing"
)

const (
	alice = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
	bob   = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
)

func testJournal(t *testing.T) *server {
	t.Helper()

	root := t.TempDir()
	held := &store{root: root}
	if err := held.ensure(); err != nil {
		t.Fatal(err)
	}

	return &server{
		store:  held,
		index:  &index{root: root, log: os.Stderr},
		pusher: &pusher{store: held, log: os.Stderr},
	}
}

func run(t *testing.T, s *server, from, message string) string {
	t.Helper()

	out, err := s.run(from, message)
	if err != nil {
		t.Fatalf("%q: %v", message, err)
	}
	return out
}

func fails(t *testing.T, s *server, from, message, want string) {
	t.Helper()

	_, err := s.run(from, message)
	if err == nil {
		t.Fatalf("%q passed", message)
	}
	if !strings.HasPrefix(err.Error(), want) {
		t.Errorf("%q gave %q, want %q", message, err, want)
	}
}

// rev reads the revision out of a reply.
func rev(reply string) string {
	line := strings.SplitN(reply, "\n", 2)[0]
	return strings.TrimPrefix(line, "rev: ")
}

func TestWriteReadAndList(t *testing.T) {
	s := testJournal(t)

	written := run(t, s, alice, `write hrs/ablations/lr --title "LR sweep" --tags a,b`+"\n# LR sweep\n\nTried 3e-4.\n")
	if !strings.HasPrefix(written, "rev: ") {
		t.Fatalf("write = %q", written)
	}

	read := run(t, s, alice, "read hrs/ablations/lr")
	for _, want := range []string{"title: \"LR sweep\"", "author: " + alice, "# LR sweep", "Tried 3e-4."} {
		if !strings.Contains(read, want) {
			t.Errorf("read does not hold %q:\n%s", want, read)
		}
	}

	if got := run(t, s, alice, "ls"); got != "hrs" {
		t.Errorf("ls = %q", got)
	}
	if got := run(t, s, alice, "ls hrs"); got != "hrs/ablations" {
		t.Errorf("ls hrs = %q", got)
	}
	if got := run(t, s, alice, "ls hrs/ablations"); got != "hrs/ablations/lr\tLR sweep" {
		t.Errorf("ls hrs/ablations = %q", got)
	}
}

func TestAWriteThatMovedIsRefused(t *testing.T) {
	s := testJournal(t)

	first := rev(run(t, s, alice, "write hrs/a/p --body \"one\""))
	run(t, s, alice, "write hrs/a/p --body \"two\"")

	fails(t, s, alice, `write hrs/a/p --if-rev `+first+` --body "three"`, "conflict")

	current := rev(run(t, s, alice, "read hrs/a/p"))
	if _, err := s.run(alice, `write hrs/a/p --if-rev `+current+` --body "three"`); err != nil {
		t.Errorf("a write at the current revision failed: %v", err)
	}
}

func TestAWriteNeedsOneBody(t *testing.T) {
	s := testJournal(t)

	fails(t, s, alice, "write hrs/a/p", "missing body")
	fails(t, s, alice, "write hrs/a/p --body \"one\"\ntwo", "invalid_arguments")
}

// The body keeps its newlines, its quotes, and a word that looks like a flag.
func TestABodyKeepsWhatItHolds(t *testing.T) {
	s := testJournal(t)

	body := "line one\nline two --not-a-flag \"quoted\"\n"
	run(t, s, alice, `write hrs/a/p --body "line one\nline two --not-a-flag \"quoted\""`)

	read := run(t, s, alice, "read hrs/a/p")
	if !strings.HasSuffix(read, body) {
		t.Errorf("read = %q, want it to end with %q", read, body)
	}
}

func TestTheProjectBelongsToItsCreator(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, "write hrs/a/p --body \"one\"")

	fails(t, s, bob, "read hrs/a/p", "forbidden")
	fails(t, s, bob, "write hrs/a/p --body \"two\"", "forbidden")
	fails(t, s, bob, "acl hrs add "+bob, "forbidden")

	if got := run(t, s, alice, "acl hrs ls"); got != alice {
		t.Errorf("the list holds %q", got)
	}

	run(t, s, alice, "acl hrs add "+bob)
	if _, err := s.run(bob, "read hrs/a/p"); err != nil {
		t.Errorf("the citizen that was added cannot read: %v", err)
	}

	// The owner never leaves the list.
	run(t, s, alice, "acl hrs rm "+alice)
	if got := run(t, s, alice, "acl hrs ls"); !strings.HasPrefix(got, alice) {
		t.Errorf("the owner left the list: %q", got)
	}

	run(t, s, alice, "acl hrs rm "+bob)
	fails(t, s, bob, "read hrs/a/p", "forbidden")

	fails(t, s, alice, "acl hrs add short", "invalid_pubkey")
	fails(t, s, alice, "acl missing ls", "forbidden")
}

func TestAppendEditAndHistory(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, "write hrs/a/p --body \"first\"")

	// The text of an append ends at the first flag, so a word that looks
	// like one travels as a JSON string, as the CLI writes it.
	run(t, s, alice, `append hrs/a/p "a line with --dashes inside\nand another"`)
	read := run(t, s, alice, "read hrs/a/p")
	if !strings.Contains(read, "first\n\na line with --dashes inside\nand another\n") {
		t.Errorf("read = %q", read)
	}

	current := rev(read)
	run(t, s, alice, `edit hrs/a/p --if-rev `+current+` --find first --replace second`)
	if read := run(t, s, alice, "read hrs/a/p"); !strings.Contains(read, "second") {
		t.Errorf("the edit did not land: %q", read)
	}

	fails(t, s, alice, "edit hrs/a/p --find x", "missing --if-rev")
	fails(t, s, alice, "edit hrs/a/p --if-rev "+rev(run(t, s, alice, "read hrs/a/p")), "missing --find")

	history := run(t, s, alice, "history hrs/a/p")
	if len(strings.Split(history, "\n")) != 3 {
		t.Errorf("the history holds %d lines:\n%s", len(strings.Split(history, "\n")), history)
	}
	fails(t, s, alice, "history hrs/a/missing", "not_found")
}

func TestTheNumbersOfANotebook(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, "write hrs/a/p --body \"one\"")

	out := run(t, s, alice, "kpi set hrs/a auc 0.871 --ref df62851")
	if !strings.Contains(out, "auc=0.871") || !strings.Contains(out, "ref=df62851") {
		t.Errorf("kpi set = %q", out)
	}

	run(t, s, alice, "kpi set hrs/a auc 0.9")
	run(t, s, alice, "kpi set hrs/a steps 1000")

	log := run(t, s, alice, "kpi log hrs/a auc")
	if len(strings.Split(log, "\n")) != 2 {
		t.Errorf("the log holds %q", log)
	}

	latest := run(t, s, alice, "kpi latest hrs/a")
	rows := strings.Split(latest, "\n")
	if len(rows) != 2 || !strings.Contains(rows[0], "auc=0.9") || !strings.Contains(rows[1], "steps=1000") {
		t.Errorf("the latest hold %q", latest)
	}

	fails(t, s, alice, "kpi set hrs/a auc not-a-number", "invalid_number")
	if got := run(t, s, alice, "kpi latest hrs/missing"); got != "no records" {
		t.Errorf("a notebook without numbers gave %q", got)
	}
}

// A page shows the latest measurement where it marks one.
func TestAPageShowsTheNumbers(t *testing.T) {
	s := testJournal(t)

	run(t, s, alice, "write hrs/a/p --body \"auc: <!-- kpi: auc --> end\"")
	run(t, s, alice, "kpi set hrs/a auc 0.871")

	read := run(t, s, alice, "read hrs/a/p")
	if !strings.Contains(read, "auc = 0.871 (") {
		t.Errorf("read = %q", read)
	}

	// A marker without a measurement stays as it is.
	run(t, s, alice, "write hrs/a/q --body \"loss: <!-- kpi: loss --> end\"")
	if read := run(t, s, alice, "read hrs/a/q"); !strings.Contains(read, "<!-- kpi: loss -->") {
		t.Errorf("the marker went: %q", read)
	}
}

func TestAttachFetchAndLink(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, "write hrs/a/p --body \"one\"")

	data := []byte("the file bytes")
	encoded := base64.StdEncoding.EncodeToString(data)

	out := run(t, s, alice, "attach hrs/a/p --name plan.md --base64 "+encoded)
	sha := strings.TrimPrefix(strings.Split(out, "\n")[1], "sha256: ")

	if got := run(t, s, alice, "fetch "+sha); got != encoded {
		t.Errorf("fetch = %q", got)
	}

	read := run(t, s, alice, "read hrs/a/p")
	if !strings.Contains(read, "attachments:") || !strings.Contains(read, "name: plan.md") {
		t.Errorf("the page does not name the attachment:\n%s", read)
	}
	if !strings.Contains(read, "bytes: 14") {
		t.Errorf("the page does not hold the size:\n%s", read)
	}

	run(t, s, alice, "link hrs/a/p https://example.com/paper --name paper")
	read = run(t, s, alice, "read hrs/a/p")
	if !strings.Contains(read, "links:") || !strings.Contains(read, "uri: https://example.com/paper") {
		t.Errorf("the page does not hold the link:\n%s", read)
	}

	// A file on the machine of the caller names that caller.
	run(t, s, alice, "link hrs/a/p file:///tmp/data.csv")
	if read := run(t, s, alice, "read hrs/a/p"); !strings.Contains(read, "host: "+alice) {
		t.Errorf("the link does not name the host:\n%s", read)
	}

	fails(t, s, alice, "link hrs/a/p not-a-uri", "invalid_uri")
	fails(t, s, alice, "attach hrs/a/p --base64 "+encoded, "missing --name")
	fails(t, s, alice, "attach hrs/a/p --name x --base64 not-base64!", "invalid_base64")
	fails(t, s, alice, "fetch short", "not_found")
}

func TestReadsARangeOfLines(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, `write hrs/a/p --body "one\ntwo\nthree\nfour"`)

	// The range counts the lines of the page, and the revision stands above
	// them.
	got := run(t, s, alice, "read hrs/a/p --lines 6:7")
	if got != "rev: "+rev(got)+"\none\ntwo" {
		t.Errorf("the range gave %q", got)
	}
}

func TestRefusesAnAddressThatIsNotOne(t *testing.T) {
	s := testJournal(t)

	for _, message := range []string{
		"read hrs",
		"read hrs/a",
		"read hrs/a/p/q",
		"read HRS/a/p",
		"read ../a/p",
		"write hrs/a --body x",
		"kpi set hrs auc 1",
	} {
		fails(t, s, alice, message, "invalid_address")
	}

	fails(t, s, "", "ls", "forbidden")
	fails(t, s, alice, "nonsense", "unknown_command")

	if got := run(t, s, alice, "help"); !strings.HasPrefix(got, "journal commands") {
		t.Errorf("help = %q", got)
	}
}

// The frontmatter is written by the journal itself, so a page reads the same
// on every machine.
func TestTheFrontmatterHoldsItsOrder(t *testing.T) {
	s := testJournal(t)
	run(t, s, alice, `write hrs/a/p --title "A title" --tags one,two --body "the body"`)

	read := run(t, s, alice, "read hrs/a/p")
	body := strings.SplitN(read, "\n", 2)[1]

	if !strings.HasPrefix(body, "---\ntitle: \"A title\"\ncreated: ") {
		t.Errorf("the frontmatter starts %q", body[:60])
	}

	order := []string{"title:", "created:", "author:", "updated:", "tags:"}
	at := 0
	for _, key := range order {
		index := strings.Index(body, key)
		if index < at {
			t.Errorf("%s stands out of order", key)
		}
		at = index
	}

	if !strings.Contains(body, "tags:\n  - one\n  - two\n") {
		t.Errorf("the tags are %q", body)
	}
}
